lib/stm32/h7/pwr.c

```diff
@@ -30,15 +30,83 @@
  * You should have received a copy of the GNU Lesser General Public License
  * along with this library.  If not, see <http://www.gnu.org/licenses/>.
  */
-
+#include <libopencm3/cm3/assert.h>
+#include <libopencm3/stm32/dbgmcu.h>
 #include <libopencm3/stm32/pwr.h>
 #include <libopencm3/stm32/rcc.h>
 #include <libopencm3/stm32/syscfg.h>
 
+/* DBGMCU_IDC DEV ID values needed to account for variations between part types. */
+#define DBGMCU_IDCODE_DEV_ID_STM32H74X_5X			0x450
+#define DBGMCU_IDCODE_DEV_ID_STM32H7A3_B3_B0		0x480
 
 void pwr_set_mode_ldo(void) {
-	const uint32_t ldo_mask = (PWR_CR3_SCUEN | PWR_CR3_LDOEN | PWR_CR3_BYPASS);
-	PWR_CR3 = (PWR_CR3 & ~ldo_mask) | (PWR_CR3_SCUEN | PWR_CR3_LDOEN);
+	/* Per table in manual for SMPS, mask and set SMPSEN=0 : LDOEN=1 : BYPASS=0. */
+	const uint32_t cr3_mask = (PWR_CR3_SMPSEN | PWR_CR3_LDOEN | PWR_CR3_BYPASS);
+	PWR_CR3 = (PWR_CR3 & ~cr3_mask) | (PWR_CR3_LDOEN);
+}
+
+void pwr_set_mode_scu_ldo(void) {
+	const uint32_t cr3_mask = (PWR_CR3_SCUEN | PWR_CR3_LDOEN | PWR_CR3_BYPASS);
+	PWR_CR3 = (PWR_CR3 & ~cr3_mask) | (PWR_CR3_SCUEN | PWR_CR3_LDOEN);
+}
+
+void pwr_set_mode_smps_ldo(bool supply_external, uint32_t smps_level, bool use_ldo) {
+	uint32_t cr3_mask, cr3_set;
+	cr3_mask = (PWR_CR3_SMPSEXTHP | PWR_CR3_SMPSEN | PWR_CR3_LDOEN | PWR_CR3_BYPASS);
+	cr3_mask |= PWR_CR3_SMPSLEVEL_MASK << PWR_CR3_SMPSLEVEL_SHIFT;
+
+	/* Default, take in unconditional settings, will OR in the rest. */
+	cr3_set = PWR_CR3_SMPSEN | (smps_level << PWR_CR3_SMPSLEVEL_SHIFT);
+	if (supply_external) {
+		cm3_assert(smps_level != PWR_CR3_SMPSLEVEL_VOS);  	/* Unsupported setting! */
+		cr3_set |= PWR_CR3_SMPSEXTHP;
+	}
+
+	if (use_ldo) {
+		cr3_set |= PWR_CR3_LDOEN;
+	}
+	PWR_CR3 = (PWR_CR3 & ~cr3_mask) | cr3_set;
+}
+
+void pwr_set_mode_bypass(void) {
+	const uint32_t cr3_mask = (PWR_CR3_SMPSEN | PWR_CR3_LDOEN | PWR_CR3_BYPASS);
+	PWR_CR3 = (PWR_CR3 & ~cr3_mask) | PWR_CR3_BYPASS;
+}
+
+void pwr_set_mode_scu_bypass(void) {
+	const uint32_t cr3_mask = (PWR_CR3_SCUEN | PWR_CR3_LDOEN | PWR_CR3_BYPASS);
+	PWR_CR3 = (PWR_CR3 & ~cr3_mask) | (PWR_CR3_SCUEN | PWR_CR3_BYPASS);
+}
+
+
+void pwr_set_mode(enum pwr_sys_mode mode, uint8_t smps_level) {
+	switch (mode) {
+		uint32_t reg_level;
+		case PWR_SYS_SCU_LDO:
+			pwr_set_mode_scu_ldo();
+			break;
+		case PWR_SYS_SCU_BYPASS:
+			pwr_set_mode_scu_bypass();
+			break;
+		case PWR_SYS_LDO:
+			pwr_set_mode_ldo();
+			break;
+		case PWR_SYS_SMPS_DIRECT:
+		case PWR_SYS_SMPS_LDO:
+			reg_level = PWR_CR3_SMPSLEVEL_VOS << PWR_CR3_SMPSLEVEL_SHIFT;
```

Comment:
and you've shifted it again here.  (and again on line 102)  Can you not just drop the intermediary variable?

Reply:
Nice catch! Will remove the redundant shift bug.